### ISSUE TITLE
Use errors codes for fastify errors

### ISSUE THREAD

**Issue by StarpTech** [fastify/fastify#909]
Hi,
when a user wants to customize the error response he must to do something like this:

```js
app.setErrorHandler(function (error, request, reply) {
  if (error.validation) {
    reply.code(400).send(error.validation)
  } else {
    reply.send(error)
  }
})
```

`error.validation` indicates the ajv error

it would be much cleaner to use error codes to avoid collisions and be more expressive.

```js
app.setErrorHandler(function (error, request, reply) {
  if (error.code === fastify.errors.ERR_VALIDATION) {
    reply.code(400).send(error.validation)
  } else {
    reply.send(error)
  }
})
```

We could add following error-codes:

* `ERR_VALIDATION`: Internal validation lib [returns an error](https://github.com/fastify/fastify/blob/44b68625efdfc08e1977664bbc589d434301b4eb/lib/validation.js#L102)
* `ERR_EOS`: error occurred during [stream handling]
(https://github.com/fastify/fastify/blob/master/lib/reply.js#L221)

**Comment by delvedor** [comment]
I agree that have some well defined error codes could be a nice feature to have.

**Comment by mcollina** [comment]
Definitely yes. However we should think this through very well.

The code should be namespaced, something like `ERR_FST_REQUEST_VALIDATION`, or ` FST_ERR_`. Something that identify this error code univocally (`ERR_` prefix is used by node).

We plan to write a userland implementation of the error code system that we have in core for readable-stream@3. I think we should use that here too. This might take a while, 👍 if you want to go ahead anyway.

**Comment by StarpTech** [comment]
> The code should be namespaced, something like ERR_FST_REQUEST_VALIDATION, or  FST_ERR_.

👍 Agree `FST_ERR_*` is great

**Comment by StarpTech** [comment]
I will prepare a PR.

**Comment by andrewzey** [comment]
Any progress on this? It would be great to be able to handle internal fastify errors separately.

**Comment by cemremengu** [comment]
@andrewzey yes very soon :) see #1168
